pkg/util/volumeclaimtemplate/volume_templates_hash.go

```diff
@@ -0,0 +1,136 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	hashutil "k8s.io/kubernetes/pkg/util/hash"
+)
+
+var (
+
+	// VolumeClaimTemplateHashAnnotation represents the specs of volumeclaimtemplates hash
+	VolumeClaimTemplateHashAnnotation = "kruise.io/cloneset-volumeclaimtemplate-hash"
+)
+
+type VolumeClaimTemplatesHasher interface {
+	// GetExpectHash get hash from volume claim templates
+	GetExpectHash(templates []v1.PersistentVolumeClaim) uint64
+	// GetCurrentHash get hash from pvcs current used when the feature first effective or hash in annotation being deleted
+	// deprecated: can not cover any scenario, some pvc type will fill some fields in pvc field however not found in vcTemplate
+	GetCurrentHash(volumes []v1.Volume, getter PVCGetter) (uint64, error)
+}
+
+type PVCGetter = func(string) *v1.PersistentVolumeClaim
+
+type volumeClaimTemplatesHasher struct{}
+
+func NewVolumeClaimTemplatesHasher() VolumeClaimTemplatesHasher {
+	return &volumeClaimTemplatesHasher{}
+}
+
+func (h *volumeClaimTemplatesHasher) GetExpectHash(templates []v1.PersistentVolumeClaim) uint64 {
+	var pvcs []v1.PersistentVolumeClaim
+	for i := range templates {
+
+		pvcs = append(pvcs, v1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: templates[i].Name,
+			},
+			Spec: templates[i].Spec,
+		})
+	}
+
+	sort.SliceStable(pvcs, func(i, j int) bool {
+		return pvcs[i].Name < pvcs[j].Name
+	})
+	return hashVolumeClaimTemplate(pvcs)
+}
+
+func (h *volumeClaimTemplatesHasher) GetCurrentHash(volumes []v1.Volume, getter PVCGetter) (uint64, error) {
+	var pvcs []v1.PersistentVolumeClaim
+	for i := range volumes {
+		if volumes[i].PersistentVolumeClaim == nil {
+			continue
+		}
+		pvcName := volumes[i].PersistentVolumeClaim.ClaimName
+		pvc := getter(volumes[i].PersistentVolumeClaim.ClaimName)
+		if pvc == nil {
+			// error here
+			return 0, fmt.Errorf("pvc %v not exist", pvcName)
+		}
+
+		pvcs = append(pvcs, v1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: volumes[i].Name,
+			},
+			Spec: pvc.Spec,
+		})
+	}
+
+	sort.SliceStable(pvcs, func(i, j int) bool {
+		return pvcs[i].Name < pvcs[j].Name
+	})
+	return hashVolumeClaimTemplate(pvcs), nil
+}
+
+func hashVolumeClaimTemplate(templates []v1.PersistentVolumeClaim) uint64 {
+	hash := fnv.New32a()
+	envsJSON, _ := json.Marshal(templates)
+	hashutil.DeepHashObject(hash, envsJSON)
+	return uint64(hash.Sum32())
+}
+
+func GetVCTemplatesHash(revision *apps.ControllerRevision) (string, bool) {
+	if len(revision.Annotations) > 0 {
+		val, exist := revision.Annotations[VolumeClaimTemplateHashAnnotation]
+		return val, exist
+	}
+	return "", false
+}
+
+func PatchVCTemplateHash(revision *apps.ControllerRevision, vcTemplates []v1.PersistentVolumeClaim) {
+	if len(revision.Annotations) == 0 {
+		revision.Annotations = make(map[string]string)
+	}
+	if len(vcTemplates) != 0 {
+		// get hash of vct
+		vcTemplateHash := NewVolumeClaimTemplatesHasher().GetExpectHash(vcTemplates)
```

Comment:
It's not necessary to initialize an object for each call, is it better to have a global variable?